src/patterns/patterns.py

```diff
@@ -599,3 +599,286 @@ def make_file_developer_df(
             sorted_hot_files = hot_files[sorted_dev_list]
         self.top_N_map = sorted_hot_files
         return sorted_hot_files, stats_df
+
+    def extract_directories(self):
+        print("INFO: Extracting head directories from filepaths...")
+        extracted_col = self.commit_data["filepath"]
+        df = pd.DataFrame(extracted_col)
+        filepaths = df["filepath"].tolist()
+        
+        # extracting the name of head directory from each filepath
+        for i in range(len(filepaths)):
+            temp = filepaths[i]
+            if(filepaths[i].find('src/') != -1):
+                split = temp.split("src/", 1)
+                temp = split[1]
+                if(temp.find('/') != -1):
+                    split = temp.split("/", 1)
+                    temp = split[0]
+                filepaths[i] = temp
+            elif(filepaths[i].find('/') != -1):
+                split = temp.split("/", 1)
+                temp = split[0]
+            filepaths[i] = temp
```

Comment:
Isn't there a more concise way of doing this, e.g., with `os.path` methods?

Reply:
I am making some other changes in this part and will see if this can be changed while doing that. Will come as part of another pull request.